src/org/rascalmpl/util/maven/Artifact.java

```diff
@@ -181,109 +181,108 @@ public boolean equals(Object obj) {
-        Set<WithoutVersion> excludes, ArrayList<Artifact> result, MavenParser parser) {
-        var nextLevel = new ArrayList<ResolveState>(dependencies.size());
-        for (var d : dependencies) {
-            var coordinate = d.getCoordinate();
+    private static void calculateClassPath(Scope forScope, Queue<ResolveState> resolveQueue, ArrayList<Artifact> result, MavenParser parser) {
+        Set<ResolveKey> alreadyResolved = new HashSet<>();
+        Map<WithoutVersion, String> resolvedVersions = new HashMap<>();
+        var rangedDeps = new HashMap<WithoutVersion, SortedSet<String>>();
 
-            boolean alreadyResolved;
+        while (!resolveQueue.isEmpty()) {
+            var state = resolveQueue.poll();
+            var artifact = state.artifact;
+            for (var d : artifact.dependencies) {
+                var coordinate = d.getCoordinate();
+
+                var versionLess = coordinate.versionLess();
+
+                var version = coordinate.getVersion();
+                if (isVersionRange(version)) {
+                    try {
+                        version = artifact.ourResolver.findLatestMatchingVersion(coordinate.getGroupId(), coordinate.getArtifactId(), version);
+                        coordinate = new ArtifactCoordinate(coordinate.getGroupId(), coordinate.getArtifactId(), version, coordinate.getClassifier());
+                    }
+                    catch (UnresolvableModelException e) {
+                        artifact.messages = artifact.messages.append(Messages.error("Version range error: " + e.getMessage(), d.getPomLocation()));
+                        continue;
+                    }
+
+                    var versions = rangedDeps.computeIfAbsent(versionLess, k -> new TreeSet<>());
+                    if (versions.add(version) && versions.size() == 2) { // Only add a warning once
+                        String effectiveVersion = versions.first();
+                        artifact.messages = artifact.messages.append(Messages.warning("Multiple version ranges found for " + versionLess + " are used. "
+                            + "It is better to lock the desired version in your own pom to a specifick version, for example <version>["
+                            + effectiveVersion + "]</version>", d.getPomLocation()));
+                    }
+                }
 
-            var versionLess = coordinate.versionLess();
+                String resolvedVersion = resolvedVersions.get(versionLess);
 
-            var version = coordinate.getVersion();
-            if (isVersionRange(version)) {
-                try {
-                    version = ourResolver.findLatestMatchingVersion(coordinate.getGroupId(), coordinate.getArtifactId(), version);
-                    coordinate = new ArtifactCoordinate(coordinate.getGroupId(), coordinate.getArtifactId(), version, coordinate.getClassifier());
+                boolean resolved;
```

Comment:
how about: `var resolved = resolvedVersion != null;` and then use that boolean in the if below?